src/Common/src/System/Globalization/IdnMapping.Unix.cs

```diff
@@ -36,25 +54,43 @@ private unsafe string GetAsciiCore(char* unicode, int count)
         private unsafe string GetUnicodeCore(char* ascii, int count)
         {
             uint flags = Flags;
-            CheckInvalidIdnCharacters(ascii, count, flags, "ascii");
+            CheckInvalidIdnCharacters(ascii, count, flags, nameof(ascii));
 
-            char[] buf = new char[count];
-
-            for (int attempts = 2; attempts > 0; attempts--)
+            const int StackAllocThreshold = 512;
+            if (count < StackAllocThreshold)
             {
-                int realLen = Interop.GlobalizationNative.ToUnicode(flags, ascii, count, buf, buf.Length);
-
-                if (realLen == 0)
+                char* output = stackalloc char[count];
+                return GetUnicodeCore(ascii, count, flags, output, count, reattempt: true);
+            }
+            else
+            {
+                char[] output = new char[count];
+                fixed (char* pOutput = output)
                 {
-                    break;
+                    return GetUnicodeCore(ascii, count, flags, pOutput, count, reattempt: true);
                 }
+            }
+        }
+
+        private unsafe string GetUnicodeCore(char* ascii, int count, uint flags, char* output, int outputLength, bool reattempt)
+        {
+            int realLen = Interop.GlobalizationNative.ToUnicode(flags, ascii, count, output, outputLength);
 
-                if (realLen <= buf.Length)
+            if (realLen == 0)
+            {
+                throw new ArgumentException(SR.Argument_IdnIllegalName, nameof(ascii));
```

Comment:
if reading realLen==0 and then throw, you'll know right away the behavior and don't have to read to the end to know what happen in such case. also if we add more code later under the condition (realLen > 0) that can make the code more nested. I know we don't have such nesting now but nobody knows if we'll write more code there in the future.